wagtail/contrib/settings/utils.py

```diff
@@ -0,0 +1,91 @@
+from django.conf import settings
+from django.urls import reverse
+
+from wagtail.coreutils import get_supported_content_language_variant
+from wagtail.models import Locale, TranslatableMixin
+
+
+def get_locale_for(*, request=None, model=None, instance=None):
+    """
+    Helper to retrieve the locale to use depending on the context.
+    If a locale is returned, it is annotated to indicate whether
+    it's the default language when possible.
+    """
+    if instance and not model:
+        model = type(instance)
+
+    i18n_enabled = getattr(settings, "WAGTAIL_I18N_ENABLED", False)
+    if model:
+        i18n_enabled = i18n_enabled and issubclass(model, TranslatableMixin)
+
+    if not i18n_enabled:
+        return None
+
+    locales = Locale.objects.annotate_default_language()
+
+    if instance:
+        # Check if the instance's locale was already prefetched
+        # and use it if so to avoid an additional query.
+        field = model._meta.get_field("locale")
+        if field.is_cached(instance):
+            return instance.locale
+        else:
+            return locales.get(pk=instance.locale_id)
+
+    selected_locale = request.GET.get("locale") if request is not None else None
+    if selected_locale:
+        try:
+            return locales.get(language_code=selected_locale)
+        except Locale.DoesNotExist:
+            pass
+
+    # Fall back to the default language
+    return locales.get(
+        language_code=get_supported_content_language_variant(settings.LANGUAGE_CODE)
+    )
+
+
+def is_default_locale(locale):
+    """Determines whether the given locale is the current default."""
+    # Use the `is_default_language` annotation if present.
+    if hasattr(locale, "is_default_language"):
+        return locale.is_default_language
+
+    is_default_language = locale == Locale.get_default()
+
+    # Set `is_default_language` for subsequent calls.
+    setattr(locale, "is_default_language", is_default_language)
+    return is_default_language
+
+
+def add_locale_query_string(url, locale):
+    if locale and not is_default_locale(locale):
+        url += f"?locale={locale.language_code}"
+    return url
```

Comment:
Moved `is_default_locale` to the `Locale` model.

Where would you suggest moving `get_locale_for` and `add_locale_query_string`? I thought `wagtail.coreutils`?

Reply:
@Tijani-Dia Please see my comments on the backport MR in GltLab RE these utils. I understand the aim is to pull out some repetitive code to make things in the views a little more readable (and I admire that), but I think this is the 'wrong abstraction'. We don't actually ever want to get a Locale for a `model`, because that isn't meaningful or possible. And when we do want to know the Locale for a `request`, not only does that not really need to have anything to do with `model`, but the fact that we derive it from `request.GET['locale']` is something that is really quite specific to these views, and not really something that can be reused elsewhere.

Reply:
Great shout @ababic. Always good to have a fresh pair of eyes

aside: I left the decision on this for the second review pass as it was a large enough PR to go through last time. Hope to come back this week with afresh, but it didn't happen.

Reply:
@ababic I mainly copied the logic in the `LocaleMixin` here: https://github.com/wagtail/wagtail/blob/main/wagtail/admin/views/generic/mixins.py#L87, it checks for both the model and the `locale` parameter in the `GET` request.

Reply:
@Tijani-Dia I see. From a design perpective, this works as a view method because all of the data is accessible from the current view instance. However, it doesn't translate very well to a separate function (where everything must be provided as an argument). Do you see what I mean? It's actually adding unnecessary complication in a couple of cases.